components/server/src/authorization/spicedb.ts

```diff
@@ -36,49 +66,106 @@ export function spiceDBConfigFromEnv(): SpiceDBClientConfig | undefined {
 
 export class SpiceDBClientProvider {
     private client: Client | undefined;
+    private previousClientOptionsString: string = DEFAULT_FEATURE_FLAG_VALUE;
+    private clientOptions: grpc.ClientOptions;
 
     constructor(
         private readonly clientConfig: SpiceDBClientConfig,
         private readonly interceptors: grpc.Interceptor[] = [],
-    ) {}
+    ) {
+        this.clientOptions = DefaultClientOptions;
+        this.reconcileClientOptions()
+            .then(() => {})
+            .catch((e) => {
+                log.error("[spicedb] Failed to reconcile client options", e);
+            });
+    }
 
-    getClient(): SpiceDBClient {
-        if (!this.client) {
-            this.client = v1.NewClient(
+    private async reconcileClientOptions() {
+        const doReconcileClientOptions = async () => {
+            try {
+                const customClientOptions = await getExperimentsClientForBackend().getValueAsync(
+                    "spicedb_client_options",
+                    DEFAULT_FEATURE_FLAG_VALUE,
+                    {},
+                );
+                if (customClientOptions === this.previousClientOptionsString) {
+                    return;
+                }
+                let clientOptions = DefaultClientOptions;
+                if (customClientOptions && customClientOptions != DEFAULT_FEATURE_FLAG_VALUE) {
+                    clientOptions = JSON.parse(customClientOptions);
+                }
+                if (this.client != null) {
+                    const newClient = this.createClient(clientOptions);
+                    const oldClient = this.client;
+                    this.client = newClient;
+
+                    log.info("[spicedb] Client options changes", {
+                        clientOptions: new TrustedValue(clientOptions),
+                    });
+
+                    // close client after 10s to make sure most pending requests on the previous client are finished.
+                    setTimeout(() => {
+                        this.closeClient(oldClient);
+                    }, 10 * 1000);
+                }
+                this.clientOptions = clientOptions;
+                // `createClient` will use the `DefaultClientOptions` to create client if the value on Feature Flag is not able to create a client
+                // but we will still write `previousClientOptionsString` here to prevent retry loops.
+                this.previousClientOptionsString = customClientOptions;
+            } catch (e) {
+                log.error("[spicedb] Failed to parse custom client options", e);
+            }
+        };
+        while (true) {
```

Comment:
Why not wrap this in an `async () => { ... }()`.
And have the loop look like:
```
while (true) {
    try {
        doReconcileOptions(); // all stuff without try...catch
    } catch (err) {
        log.error("[spicedb] Failed to parse custom client options", e);
    }
    await new Promise((resolve) => setTimeout(resolve, 60 * 1000));
}
```

This would make it way easier to reason a) what the code is doing and b) that it's correct (as in: does not accidentally bail out, always respects timeout, etc.)

Reply:
> reconcileClientOptions is not expected to throw any error, or wait on anything. It should be sync and void.

These two comments are conflicted?